source/interfaces/validator-function.ts

```diff
@@ -0,0 +1,21 @@
+module fjs {
+
+  /**
+   * Custom validation function signature.
+   *
+   * <p>Validations functions are provided two parameters:
+   * <ol>
+   *   <li>The value of the field/attribute.
+   *       This is often the only piece of data needed for a custom validation.
+   *   <li>The form data object (all fields)
+   *       This additional data may be needed when a field depends on another field (e.g. confirm password).
+   * </ol>
+   *
+   * <p>Custom validators are responsible for returning either a truthy/falsy value OR a Promise.
+   * Synchronous validations can return TRUE to indicate a valid value or FALSE to indicate an invalid value.
+   * Use a Promise to perform asynchronous validation or to return a custom failure message.
+   */
+  export interface ValidatorFunction {
+    (value:any, formData:any):boolean|Promise<string>;
```

Comment:
> Shouldn't string be a valid return type

Potentially. What would it mean for a custom validator to return a string? Does that indicate a valid or invalid value? I *assume* a string would have to be treated as an error message but I worry this confuses our truthy/falsy return scenario. Do we really need it? If you want to return a synchronous failure message, you could always return `Promise.reject("Your message")`.

> Promise is probably resolvable to boolean too

I think this would be redundant. A resolved promise means a valid value; a rejected promise means an invalid value. The only reason a `string` type is specified is to support the case of an overriding, custom validation failure message.